Add new font selection APIs to DWrite API spec (#1967)
microsoft/WindowsAppSDK: specs/dwritecore/DWriteCore.md

New APIs introduced by DWriteCore include:

 - The exported function is renamed `DWriteCoreCreateFactory` to avoid a naming collision with the
   `DWriteCreateFactory` function exported by the system version of DirectWrite (DWrite.dll)

>  - A new `DWRITE_FACTORY_TYPE` enumerator for creating an isolated factory object

Reviewer: The constants in an enumeration are called enumerators.

See also:
- https://docs.microsoft.com/en-us/cpp/cpp/enumerations-cpp?view=msvc-170
- https://en.cppreference.com/w/cpp/language/enum

>  - A new `DWRITE_FACTORY_TYPE` enumerator for creating an isolated factory object

Reviewer 2: RECOMMEND: The name of the thing is an enumeration, which is a set of named/grouped constants.

>  - A new `DWRITE_FACTORY_TYPE` enumerator for creating an isolated factory object

Reviewer: Is there any basis for `enumerant`? Making up words that aren't used in the general community is unhelpful for a public specification. Perhaps this can be refactored out as such:

```suggestion
 - Adds `DWRITE_FACTORY_TYPE` to create an isolated factory object
```

>  - A new `DWRITE_FACTORY_TYPE` enumerator for creating an isolated factory object

Reviewer 3: [The Windows Runtime calls them "values"](https://github.com/microsoft/xlang/blob/master/docs/xplatwinrt/XSPEC02%20-%20Metadata%20Representation.md):

> After the enum value definition comes a field definition for each of the values in the enumeration

That seems like a word whose meaning is clear to all readers, despite perhaps not matching what C++ calls it (because the C++ name is easily confused with C# enumerators).

>A new `DWRITE_FACTORY_TYPE` value which specifies...

>  - A new `DWRITE_FACTORY_TYPE` enumerator for creating an isolated factory object

Reviewer: @oldnewthing <shakes hand> I like it; agreed!

>  - A new `DWRITE_FACTORY_TYPE` enumerator for creating an isolated factory object

Author: Sounds good. I'm going with:
A new `DWRITE_FACTORY_TYPE` value which specifies an isolated factory object

 - A new method for getting pixel data from a bitmap render target
 - New methods for selecting fonts from a font set
 - A new factory method to release unreferenced fonts

# Factory Creation

The new `DWRITE_FACTORY_TYPE_ISOLATED2` enumerator complements the existing factory types. It is similar to
`DWRITE_FACTORY_TYPE_ISOLATED`, but is more locked down in two ways. First, it only caches data in-process, and
will not attempt to either read from or write to a cross-process font cache or persisted cache file. Second,

> a restricted factory only enumerates well-known fonts in the system font collection. The well-known-fonts

Reviewer 2: RECOMMEND: Yes, be consistent on "restricted" vs "isolated"

> a restricted factory only enumerates well-known fonts in the system font collection. The well-known-fonts

Author: This sentence applies only to an isolated2 factory, so I'm going with "a `DWRITE_FACTORY_TYPE_ISOLATED2` factory only enumerates..."

font in a WWS family must have a unique location defined by its coordinates along those three axes. To
select a font, one specifies a WWS family name and weight, stretch, and style parameters.

By contrast, the typographic font family model has an N-dimensional design space. A font designer can 
define any number of design axes, each identified by a four-character _axis tag_. A given font’s location 
in the N-dimensional design space is defined by an array of _axis values_, where each axis value comprises 
an axis tag and a floating-point value. To select a font, one specifies a typographic family name and an 
array of axis values (`DWRITE_FONT_AXIS_VALUE` structures).

Although the number of font axes is open-ended, there are a few registered axes with standard meanings,
and the weight, stretch, and style values can be mapped to registered axis values. `DWRITE_FONT_WEIGHT`
can be mapped to a "wght" (`DWRITE_FONT_AXIS_TAG_WEIGHT`) axis value. `DWRITE_FONT_STRETCH` can be mapped
to a "wdth" (`DWRITE_FONT_AXIS_TAG_WIDTH`) axis value. `DWRITE_FONT_STYLE` can be mapped to a combination

> of "ital" and "slnt" (`DWRITE_FONT_AXIS_TAG_ITALIC` and `DWRITE_FONT_AXIS_TAG_SLANT`) axis values. 

Reviewer 2: DISCUSSION: They're proper dimensions in the OpenType spec. They're generally basically the same, but in some fonts there is additional shaping for "slant" vs "italic."

> of "ital" and "slnt" (`DWRITE_FONT_AXIS_TAG_ITALIC` and `DWRITE_FONT_AXIS_TAG_SLANT`) axis values. 

Reviewer 4: `ital` is a more discrete 0/1 concept (even if some variable fonts technically let you interpolate between them), whereas `slnt` allows finer control over the exact angle, enabling light right slant, far right slant, and even left leaning if that is desired. e.g. 
![image](https://user-images.githubusercontent.com/1809166/150017055-374b51a8-bc80-431b-8a27-1c0753210de8.png)

So they are related but not *quite* the same, as some font are slanted (purely oblique shear) but not "italic" (which modifies the letter forms too), and some very rare fonts (just to throw a monkey wrench into a simple story) can appear italic but upright 🙃.
![image](https://user-images.githubusercontent.com/1809166/150021919-b9df1e87-bf1d-4bfb-a1b3-c26e0cff26e3.png)

(but yes, in the general case, one implies the other)

}
```

## Using Font Selection APIs

This section shows a complete console application that demonstrates the `IDWriteFontSet4::GetMatchingFonts`
and `IDWriteFontSet4::GetDerivedFontAxisValues` methods. First let's include some headers:

```c++
#include <dwrite_core.h>
#include <wil/com.h>
#include <iostream>
#include <string>
#include <vector>
```

The `IDWriteFontSet4::GetMatchingFonts` method returns a list of fonts in priority order that match the
specified family name and axis values. The following `MatchAxisValues` function outputs the parameters 
to `GetMatchingFonts` and the list of matching fonts in the returned font set.

```c++
// Forward declarations of overloaded output operators used by MatchAxisValues.
std::wostream& operator<<(std::wostream& out, DWRITE_FONT_AXIS_VALUE axisValue);
std::wostream& operator<<(std::wostream& out, IDWriteFontFile& fileReference);

> std::wostream& operator<<(std::wostream& out, IDWriteFontFaceReference1& faceReference);

Reviewer 2: DISCUSSION: These are Nano-COM references, where other code would have said IFoo*, this uses IFoo&

> std::wostream& operator<<(std::wostream& out, IDWriteFontFaceReference1& faceReference);

Author: I'll change these to take pointers instead of references. I agree a reference to a nano-COM interface is weird, as is passing a non-const reference to an operator.

//
// MatchAxisValues calls IDWriteFontSet4::GetMatchingFonts with the
// specific parameters and outputs the matching fonts.
//
void MatchAxisValues(
    IDWriteFontSet4* fontSet,
    _In_z_ WCHAR const* familyName,
    std::vector<DWRITE_FONT_AXIS_VALUE> const& axisValues,
    DWRITE_FONT_SIMULATIONS allowedSimulations
    )
{
    // Write the input parameters.
    std::wcout << L"GetMatchingFonts(\"" << familyName << L"\", {";

>     for (DWRITE_FONT_AXIS_VALUE axisValue : axisValues)

Reviewer 2: SAMPLE: Consider fixing

>     for (DWRITE_FONT_AXIS_VALUE axisValue : axisValues)

Reviewer 5: @niklasb-ms , did you want to update this as well?

    FontStyleParams(float fontSize) : fontSize{ fontSize } {}

    DWRITE_FONT_WEIGHT fontWeight = DWRITE_FONT_WEIGHT_NORMAL;
    DWRITE_FONT_STRETCH fontStretch = DWRITE_FONT_STRETCH_NORMAL;
    DWRITE_FONT_STYLE fontStyle = DWRITE_FONT_STYLE_NORMAL;
    float fontSize = 0.0f;
};

//
// MatchFont calls IDWriteFontSet4::GetDerivedFontAxisValues to map the input
// paramters to axis values and then calls MatchAxisValues.
//
void MatchFont(
    IDWriteFactory7* factory,
    _In_z_ WCHAR const* familyName,
    FontStyleParams styleParams = {},
    std::vector<DWRITE_FONT_AXIS_VALUE> axisValues = {},
    DWRITE_FONT_SIMULATIONS allowedSimulations = DWRITE_FONT_SIMULATIONS_BOLD | DWRITE_FONT_SIMULATIONS_OBLIQUE
    )
{
    wil::com_ptr<IDWriteFontSet2> fontSet2;
    THROW_IF_FAILED(factory->GetSystemFontSet(/*includeDownloadableFonts*/ false, &fontSet2));

    wil::com_ptr<IDWriteFontSet4> fontSet;
    THROW_IF_FAILED(fontSet2->QueryInterface(&fontSet));

    // Get derived axis value from the weight, stretch, style, and font size.
    // Any axes in the input vector are excluded from the output. The maximum
    // number of derived axis values is DWRITE_STANDARD_FONT_AXIS_COUNT.

>     DWRITE_FONT_AXIS_VALUE derivedAxisValues[DWRITE_STANDARD_FONT_AXIS_COUNT];

Reviewer 2: SAMPLE: Neat trick, left up to the team to apply this as they see fit.

>     DWRITE_FONT_AXIS_VALUE derivedAxisValues[DWRITE_STANDARD_FONT_AXIS_COUNT];

Author: Done.

    {
        const void* fileKey;
        UINT32 fileKeySize;
        THROW_IF_FAILED(fileReference.GetReferenceKey(&fileKey, &fileKeySize));

        WCHAR filePath[MAX_PATH];
        THROW_IF_FAILED(localLoader->GetFilePathFromKey(fileKey, fileKeySize, filePath, MAX_PATH));

        WCHAR* fileName = wcsrchr(filePath, L'\\');
        fileName = (fileName != nullptr) ? fileName + 1 : filePath;

        out << fileName;
    }
    return out;
}

// Output a font face reference.
std::wostream& operator<<(std::wostream& out, IDWriteFontFaceReference1& faceReference)
{
    // Output the font file name.
    wil::com_ptr<IDWriteFontFile> fileReference;
    THROW_IF_FAILED(faceReference.GetFontFile(&fileReference));
    std::wcout << *fileReference;

    // Output the face index if nonzero.
    UINT32 const faceIndex = faceReference.GetFontFaceIndex();
    if (faceIndex != 0)
    {
        out << L'#' << faceIndex;
    }

    // Output the axis values.
    UINT32 const axisCount = faceReference.GetFontAxisValueCount();
    std::vector<DWRITE_FONT_AXIS_VALUE> axisValues(axisCount);
    THROW_IF_FAILED(faceReference.GetFontAxisValues(axisValues.data(), axisCount));

>     for (DWRITE_FONT_AXIS_VALUE axisValue : axisValues)

Reviewer 3: ```suggestion
    for (DWRITE_FONT_AXIS_VALUE const& axisValue : axisValues)
```
avoid unnecessary copy.

Reviewer 2: SAMPLE: Fix

Author: Done.

{
    /// <summary>
    /// Computes derived font axis values from the specified font weight, stretch, style, and size.
    /// </summary>
    /// <param name="inputAxisValues">Pointer to an optional array of input axis values. Any axes present
    /// in this array are not included in the output. This is so explicit axis values take precedence over 
    /// derived axis values.</param>
    /// <param name="inputAxisCount">Size of the array of input axis values.</param>
    /// <param name="fontWeight">Font weight, used to compute "wght" axis value.</param>
    /// <param name="fontStretch">Font stretch, used to compute "wdth" axis value.</param>
    /// <param name="fontStyle">Font style, used to compute "slnt" and "ital" axis values.</param>
    /// <param name="fontSize">Font size in DIPs, used to compute "opsz" axis value. If this parameter is zero,
    /// no "opsz" axis value is added to the output array.</param>
    /// <param name="outputAxisValues">Pointer to an output array to which derived axis values are written.
    /// The size of this array must be at least DWRITE_STANDARD_FONT_AXIS_COUNT (5). The return value is 

>     /// the actual number of axis values written to this array.</param>

Reviewer 2: DISCUSSION: This is just a place to expose the method. The GetDerivedFontAxisValues is the conversion tool from the old value types to the new axis types - this method fills in the gaps of what you don't yet have. If the input set contains weight, the output set will _not_ contain weight, as it's not derived - you know it upfront.

SAMPLE: Revisit Raymond's "here's how to combine everything in a single call" model, optionally.

>     /// the actual number of axis values written to this array.</param>

Reviewer 3: Just clarifying that 5 elements is sufficient. The existing docs said that it must be "at least 5", which opened the question of "are there cases where it might write 6"? (Answer: No. It will never write 6.)

    /// <returns>Returns the actual number of derived axis values written to the output array.</returns>
    STDMETHOD_(UINT32, GetDerivedFontAxisValues)(
        _In_reads_opt_(inputAxisCount) DWRITE_FONT_AXIS_VALUE const* inputAxisValues,
        UINT32 inputAxisCount,

>         DWRITE_FONT_WEIGHT fontWeight,

Reviewer 3: How does the function report errors? It returns `UINT32`, not `HRESULT`. For example, maybe `fontWeight` is out of range. (I guess this function doesn't allocate temporary memory.)

Author: The function is no-fail. It does not allocate temporary storage. It does not care if input axis values are in range. It only looks at the input axis _tags_ to see which ones are present.

Reviewer 2: DISCUSSION: Returning UINT32 is fine here as there's no "failure" path in the algorithm.

        DWRITE_FONT_STRETCH fontStretch,
        DWRITE_FONT_STYLE fontStyle,
        float fontSize,

>         _Out_writes_to_(DWRITE_STANDARD_FONT_AXIS_COUNT, return) DWRITE_FONT_AXIS_VALUE* outputAxisValues

Reviewer 2: Is this `DWRITE_STANDARD_FONT_AXIS_COUNT` already defined somewhere?

Author: No, that is a new #define.
```
#define DWRITE_STANDARD_FONT_AXIS_COUNT 5
```
Should I give that its own heading?

Reviewer 2: DISCUSSION: Why only up to five total output values?  There are only five well-known defined output axes - weight, width, slant, italics, optical-size.  Those are the ones that can be derived from the input values passed in.

What if there's a sixth default axis?  We'd add a GetDerivedFontAxisValues2 that takes new values to do the derivation. This is for WWS->typographic migration.

CONSIDER: Should the name of this method be more "it's a bridge" like "DeriveAxisValuesFromWWSModel" or similar, or "ConvertWeightWidthStyleToFontAxis"

Author: After discussing with various people, I'm going with `DeriveAxisValuesFromWeightStretchStyle`.

Author: After further discussion, I renamed the method to ConvertWeightStretchStyleToFontAxisValues.
"FontAxisValues" instead of "AxisValues" for consistency with other DWrite methods.
"Convert" instead of "Derive" because it's more common and is already used in other DWrite methods.
